sinatra-contrib/lib/sinatra/engine_tracking.rb

```diff
@@ -29,9 +29,10 @@ def erubis?
       erb? && Tilt[:erb] == Tilt::ErubisTemplate
     end
 
-    # @return [Boolean] Returns true if current engine is `:haml`.
+    # @return [Boolean] Returns true if current engine is `:haml` and
+    # `Tilt[:haml]` is set to `Tilt::HamlTemplate`.
     def haml?
-      @current_engine == :haml
+      @current_engine == :haml && Tilt[:haml] == Tilt::HamlTemplate
```

Comment:
Why is this extra check necessary?

Reply:
`Tilt[:haml]` can also be `Hamlit::Template` so that it can be used for .haml files, and `haml?` is currently used for branches really specific for haml.gem (which Hamlit basically doesn't support), not Haml language's syntax.

Should we have the `Tilt[:haml] == Tilt::HamlTemplate` check outside `haml?` for each branches, or should we rename it to `haml_gem?`?

Reply:
Oh I understand now. I feel like this could be confusing since hamlit really is a haml engine. How many branches are there?

Reply:
There are just two actually. Does this make more sense to you then? https://github.com/sinatra/sinatra/pull/1580/commits/58e1be025826a7214dcb0ef17cb511afde84a34e

Reply:
I think I do like that better. Hopefully that helps us see areas to investigate refactoring.